pkg/rc/replication_controller.go

```diff
@@ -0,0 +1,239 @@
+package rc
+
+import (
+	"strings"
+	"time"
+
+	"github.com/square/p2/pkg/kp"
+	"github.com/square/p2/pkg/kp/rcstore"
+	"github.com/square/p2/pkg/labels"
+	"github.com/square/p2/pkg/logging"
+	"github.com/square/p2/pkg/pods"
+	"github.com/square/p2/pkg/util"
+)
+
+const (
+	rcIdLabel  = "replication_controller_id"
+	podIdLabel = "pod_id"
+)
+
+type ReplicationController interface {
+	ID() string
+	Manifest() pods.Manifest
+	NodeSelector() labels.Selector
+	PodLabels() labels.Set
+	Disabled() bool
+	DesiredReplicas() int
+	CurrentNodes() ([]string, error)
+
+	// WatchDesires() spawns a goroutine that continually watches the desired replica count
+	// of this replication controller.
+	// Upon seeing any changes, it will converge observed state to desired state.
+	// The goroutine continues running until a value is received on the quitAndAck channel.
+	// The goroutine signals it has stopped by sending back a value on quitAndAck.
+	WatchDesires(quitAndAck chan struct{}) <-chan error
+}
+
+// I just copied this from kp.Store because it was a huge interface and I really only care about these methods!
+type kpStore interface {
+	SetPod(key string, manifest pods.Manifest) (time.Duration, error)
+	DeletePod(key string) (time.Duration, error)
+}
+
+type replicationController struct {
+	rcstore.ReplicationControllerData
+
+	logger logging.Logger
+
+	kpStore       kpStore
+	scheduler     Scheduler
+	podApplicator labels.Applicator
+}
+
+// replicationController implements the ReplicationController interface.
+var _ ReplicationController = &replicationController{}
+
+func New(
+	rcData rcstore.ReplicationControllerData,
```

Comment:
It seems backwards to have `pkg/rc` require something from `pkg/kp/rcstore`, since the rc package is what should define the fields an rc needs. The `pods` package is the opposite of this.